File: src/main/java/AoC2023_16.java
import java.util.Arrays;
import java.util.List;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.Stream;

import com.github.pareronia.aoc.AssertUtils;
import com.github.pareronia.aoc.CharGrid;
import com.github.pareronia.aoc.Grid.Cell;
import com.github.pareronia.aoc.geometry.Direction;
import com.github.pareronia.aoc.geometry.Turn;
import com.github.pareronia.aoc.graph.BFS;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public final class AoC2023_16
        extends SolutionBase<AoC2023_16.Contraption, Integer, Integer> {
    
    private AoC2023_16(final boolean debug) {
        super(debug);
    }
    
    public static AoC2023_16 create() {
        return new AoC2023_16(false);
    }
    
    public static AoC2023_16 createDebug() {
        return new AoC2023_16(true);
    }
    
    @Override
    protected Contraption parseInput(final List<String> inputs) {
        return new Contraption(CharGrid.from(inputs));
    }

    @Override
    public Integer solvePart1(final Contraption contraption) {
        return contraption.getInitialEnergy();
    }
    
    @Override
    public Integer solvePart2(final Contraption contraption) {
        return contraption.getMaximalEnergy();
    }
    
    @Override
    @Samples({
        @Sample(method = "part1", input = TEST, expected = "46"),
        @Sample(method = "part2", input = TEST, expected = "51"),
    })
    public void samples() {
    }
    
    public static void main(final String[] args) throws Exception {
        AoC2023_16.create().run();
    }
    
    record Contraption(CharGrid grid) {
        
        record Beam(Cell cell, Direction dir) { }
    
        public int getInitialEnergy() {
            return getEnergised(new Beam(Cell.at(0, 0), Direction.RIGHT));
        }
        
        public int getMaximalEnergy() {
            return Stream.of(
                grid.rowIndices().intStream()
                    .mapToObj(row -> new Beam(Cell.at(row, 0), Direction.RIGHT)),
                grid.rowIndices().intStream()
                    .mapToObj(row -> new Beam(Cell.at(row, grid.getWidth() - 1), Direction.LEFT)),
                grid.colIndices().intStream()
                    .mapToObj(col -> new Beam(Cell.at(0, col), Direction.DOWN)),
                grid.colIndices().intStream()
                    .mapToObj(col -> new Beam(Cell.at(grid.getHeight() - 1, col), Direction.UP))
            ).reduce(Stream.empty(), Stream::concat)
            .mapToInt(this::getEnergised)
            .max().getAsInt();
        }

        private Stream<Beam> stream(final Beam beam, final Direction...dirs) {
            return Arrays.stream(dirs)
                .map(d -> new Beam(beam.cell.at(d), d))
                .filter(t -> grid.isInBounds(t.cell));
        }
        
        private int getEnergised(final Beam initialBeam) {
            final Function<Beam, Stream<Beam>> adjacent = beam -> {
                final char val = grid.getValue(beam.cell);
                if (val == '.'
                    || val == '|' && beam.dir.isVertical()
                    || val == '-' && beam.dir.isHorizontal()) {
                    return stream(beam, beam.dir);
                } else if (val == '/' && beam.dir.isHorizontal()) {
                    return stream(beam, beam.dir.turn(Turn.LEFT));
                } else if (val == '/' && beam.dir.isVertical()) {
                    return stream(beam, beam.dir.turn(Turn.RIGHT));
                } else if (val == '\\' && beam.dir.isHorizontal()) {
                    return stream(beam, beam.dir.turn(Turn.RIGHT));
                } else if (val == '\\' && beam.dir.isVertical()) {
                    return stream(beam, beam.dir.turn(Turn.LEFT));
                } else if (val == '|' && beam.dir.isHorizontal()) {
                    return stream(beam, Direction.UP, Direction.DOWN);
                } else if (val == '-' && beam.dir.isVertical()) {
                    return stream(beam, Direction.LEFT, Direction.RIGHT);
                }
                throw AssertUtils.unreachable();
            };

            final Set<Beam> energised = BFS.floodFill(initialBeam, adjacent);
            return (int) energised.stream().map(Beam::cell).distinct().count();
        }
    }

    private static final String TEST = """
            .|...\\....
            |.-.\\.....
            .....|-...
            ........|.
            ..........
            .........\\
            ..../.\\\\..
            .-.-/..|..
            .|....-|.\\
            ..//.|....
            """;
}

File: src/main/python/AoC2023_16.py
#! /usr/bin/env python3
#
# Advent of Code 2023 Day 16
#

import itertools
import sys
from collections.abc import Iterator
from typing import NamedTuple

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples
from aoc.geometry import Direction
from aoc.geometry import Turn
from aoc.graph import flood_fill
from aoc.grid import Cell
from aoc.grid import CharGrid

TEST = """\
.|...\\....
|.-.\\.....
.....|-...
........|.
..........
.........\\
..../.\\\\..
.-.-/..|..
.|....-|.\\
..//.|....
"""


class Beam(NamedTuple):
    cell: Cell
    dir: Direction


class Contraption(NamedTuple):
    grid: CharGrid

    def _get_energised(self, initial_beam: Beam) -> int:
        def adjacent(beam: Beam) -> Iterator[Beam]:
            val = self.grid.get_value(beam.cell)
            if (
                val == "."
                or (val == "|" and beam.dir.is_vertical)
                or (val == "-" and beam.dir.is_horizontal)
            ):
                yield Beam(beam.cell.at(beam.dir), beam.dir)
            elif val == "/" and beam.dir.is_horizontal:
                new_dir = beam.dir.turn(Turn.LEFT)
                yield Beam(beam.cell.at(new_dir), new_dir)
            elif (val == "/" and beam.dir.is_vertical) or (
                val == "\\" and beam.dir.is_horizontal
            ):
                new_dir = beam.dir.turn(Turn.RIGHT)
                yield Beam(beam.cell.at(new_dir), new_dir)
            elif val == "\\" and beam.dir.is_vertical:
                new_dir = beam.dir.turn(Turn.LEFT)
                yield Beam(beam.cell.at(new_dir), new_dir)
            elif val == "|" and beam.dir.is_horizontal:
                yield Beam(beam.cell.at(Direction.UP), Direction.UP)
                yield Beam(beam.cell.at(Direction.DOWN), Direction.DOWN)
            elif val == "-" and beam.dir.is_vertical:
                yield Beam(beam.cell.at(Direction.LEFT), Direction.LEFT)
                yield Beam(beam.cell.at(Direction.RIGHT), Direction.RIGHT)
            else:
                raise AssertionError

        energised = flood_fill(
            initial_beam,
            lambda beam: (
                b for b in adjacent(beam) if self.grid.is_in_bounds(b.cell)
            ),
        )
        return len({beam.cell for beam in energised})

    def get_initial_energy(self) -> int:
        return self._get_energised(Beam(Cell(0, 0), Direction.RIGHT))

    def get_maximal_energy(self) -> int:
        return max(
            self._get_energised(beam)
            for beam in itertools.chain(
                (
                    Beam(Cell(row, 0), Direction.RIGHT)
                    for row in range(self.grid.get_height())
                ),
                (
                    Beam(Cell(row, self.grid.get_width() - 1), Direction.LEFT)
                    for row in range(self.grid.get_height())
                ),
                (
                    Beam(Cell(0, col), Direction.DOWN)
                    for col in range(self.grid.get_width())
                ),
                (
                    Beam(Cell(self.grid.get_height() - 1, col), Direction.UP)
                    for col in range(self.grid.get_width())
                ),
            )
        )


Input = Contraption
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return Contraption(CharGrid.from_strings(list(input_data)))

    def part_1(self, contraption: Input) -> Output1:
        return contraption.get_initial_energy()

    def part_2(self, contraption: Input) -> Output2:
        return contraption.get_maximal_energy()

    @aoc_samples(
        (
            ("part_1", TEST, 46),
            ("part_2", TEST, 51),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2023, 16)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()
